src/efftester.ml

```diff
@@ -1089,6 +1152,63 @@ module GeneratorsWithContext (Ctx : Context) = struct
     in
     [ (3, gen) ]
 
+  (* FIXME: remove:
+    can goal type {!t} be (Option (typevar _)) ?
+    what if it is Option (Option (_))?
+
+  *)
+  and option_intro_rules env t eff size =
+    let gen =
+      let open Gen in
+      match t with
+      | Option t' ->
+        Gen.frequencyl [ (2, "Some"); (1, "None") ] >>= fun name ->
+        (match name with
+        | "None" -> return (Some (none t))
+        | "Some" ->
+          list_permute_term_gen_outer env t' eff (size / 3) >>= fun trm ->
+          (match trm with
+          | None -> return None
+          | Some trm' -> return (Some (some t trm')))
+        | _ -> failwith "option_intro_rules: impossible option adt_constr name")
+      | _ -> return None
+    in
+    [ ( 100,
+        fun x ->
+          (* print_endline "option_intro_rules"; *)
+          gen x )
+    ]
+
+  and option_elim_rules env t eff size =
+    let gen =
+      let open Gen in
+      StaticGenerators.basetype_gen >>= fun bt ->
+      list_permute_term_gen_outer env (Option bt) eff (size / 3) >>= function
+      | None -> return None
+      | Some match_trm ->
+        let extended_env = add_var "x" bt env in
+        list_permute_term_gen_outer extended_env t eff (size / 3) >>= function
+        | None -> return None
+        | Some some_branch_trm ->
+          list_permute_term_gen_outer env t eff (size / 3) >>= function
+          | None -> return None
+          | Some none_branch_trm ->
+            return
+              (Some
+                 (PatternMatch
+                    ( t,
+                      match_trm,
+                      [ (some t (Variable (t, "x")), some_branch_trm);
```

Comment:
With a hardcoded variable name `"x"` we can only generate 
` | Some x -> ...` clauses.
You should generate an arbitrary variable name like in `lam_rules` to have a chance for `Some a`, `Some b`, ...